src/mc_minimize/mc_minimizer_local.cu

```diff
@@ -0,0 +1,766 @@
+  int n, patch;
+  __shared__ double data[1024];
+  data[tid] = 0.0;
+  for (patch = 0; patch < number_of_patches; ++patch) {
+    n = tid + patch * 1024;
+    if (n < size)
+      data[tid] += a[n];
+  }
+  __syncthreads();
+  for (int offset = blockDim.x >> 1; offset > 0; offset >>= 1) {
+    if (tid < offset) {
+      data[tid] += data[tid + offset];
+    }
+    __syncthreads();
+  }
+  if (tid == 0)
+    *result = data[0];
+}
+
+/*
+calculate the displacement for the labelled atomsd
+*/
+static __global__ void get_displacement(
+  const Box box,
+  double* global_position,
+  double* local_position,
+  int global_N,
+  int local_N,
+  int* local_index,
+  double* outer_atoms_flags,
+  double* displacement)
+{
+  int n = blockDim.x * blockIdx.x + threadIdx.x;
+  if (n < local_N)
+  {
+    int index = local_index[n];
+    double displace = 0;
+    double r12[3];
+    for (int i = 0; i < 3; i++)
+    {
+      r12[i] = (global_position[index + i * global_N] - local_position[n + i * local_N]);
+    }
+    apply_mic(box, r12[0], r12[1], r12[2]);
+    displace = r12[0] * r12[0] + r12[1] * r12[1] + r12[2] * r12[2];
+    displacement[n] = sqrt(displace) * outer_atoms_flags[n];
+  }
+}
+
+//remove the energy of the background atoms
+static __global__ void modify_energy(
+  const int local_N,
+  double* pe,
+  double* local_sphere_flags)
+{
+  int n = blockDim.x * blockIdx.x + threadIdx.x;
+  if (n < local_N)
+  {
+    pe[n] *= local_sphere_flags[n];
+  } 
+}
+
+//calculate the max value
+__global__ void gpu_calculate_max(
+  const int size,
+  const int number_of_rounds,
+  const double* array,
+  double* max)
+{
+  const int tid = threadIdx.x;
+
+  __shared__ double s_max[1024]; // Shared memory for max values
+  s_max[tid] = 0;         
+
+  double max_value = 0;   // Initialize local max
+
+  for (int round = 0; round < number_of_rounds; ++round) {
+    const int n = tid + round * 1024;
+    if (n < size) {
+      const double f = array[n];
+      if (f > max_value)
+        max_value = f;           // Update local max
+    }
+  }
+
+  s_max[tid] = max_value;        // Write local max to shared memory
+  __syncthreads();
+
+  for (int offset = blockDim.x >> 1; offset > 0; offset >>= 1) {
+    if (tid < offset) {
+      if (s_max[tid + offset] > s_max[tid]) {
+        s_max[tid] = s_max[tid + offset];
+      }
+    }
+    __syncthreads();
+  }
+
+  if (tid == 0) {
+    max[0] = s_max[0]; // Block's final max written to global memory
+  }
+}
+
+double get_outer_average_displacement(
+  const Box box,
+  double* global_position,
+  double* local_position,
+  int global_N,
+  int local_N,
+  int* local_index,
+  double* outer_atoms_flags,
+  double* max)
+{
+  GPU_Vector<double> displacement(local_N);
+  get_displacement<<<(local_N - 1) / 64 + 1, 64>>>(
+    box,
+    global_position,
+    local_position,
+    global_N,
+    local_N,
+    local_index,
+    outer_atoms_flags,
+    displacement.data());
+  gpuDeviceSynchronize();
```

Comment:
gpuDeviceSynchronize() can slowdown code. Use this only if it is needed. Also wrap it within CHECK()